src/dune_rules/ctypes_rules.ml

```diff
@@ -0,0 +1,389 @@
+  let pr buf fmt = Printf.bprintf buf (fmt ^^ "\n") in
+  pr buf "let () =";
+  pr buf "  let concurrency = %s in" concurrency;
+  pr buf "  let prefix = Sys.argv.(2) in";
+  pr buf "  match Sys.argv.(1) with";
+  pr buf "  | \"ml\" ->";
+  pr buf "    Cstubs.write_ml ~concurrency Format.std_formatter ~prefix";
+  pr buf "      (module %s.Functions)" function_description_module;
+  pr buf "  | \"c\" ->";
+  List.iter include_headers ~f:(fun h ->
+    pr buf "    print_endline \"#include <%s>\";" h);
+  pr buf "    Cstubs.write_c ~concurrency Format.std_formatter ~prefix";
+  pr buf "      (module %s.Functions)" function_description_module;
+  pr buf "  | s -> failwith (\"unknown functions \"^s)";
+  Buffer.contents buf
+
+let write_type_gen_script ~include_headers ~dir ~filename ~sctx
+      ~type_description_module =
+  let path = Path.Build.relative dir filename in
+  let script = type_gen_gen ~include_headers ~type_description_module in
+  Super_context.add_rule ~loc:Loc.none sctx ~dir (Build.write_file path script)
+
+let write_function_gen_script ~include_headers ~sctx ~dir ~name
+      ~function_description_module =
+  let path = Path.Build.relative dir (name ^ ".ml") in
+  let script = function_gen_gen ~include_headers ~function_description_module in
+  Super_context.add_rule ~loc:Loc.none sctx ~dir (Build.write_file path script)
+
+let rule ?(deps=[]) ?stdout_to ?(args=[]) ?(targets=[]) ~exe ~sctx ~dir () =
+  let build =
+    let exe = Ok (Path.build (Path.Build.relative dir exe)) in
+    let args =
+      let targets = List.map targets ~f:(Path.Build.relative dir) in
+      let deps =
+        List.map deps ~f:(Path.relative (Path.build dir))
+        |> Dep.Set.of_files
+      in
+      let open Command.Args in
+      [ Hidden_targets targets
+      ; Hidden_deps deps
+      ; As args ]
+    in
+    let stdout_to = Option.map stdout_to ~f:(Path.Build.relative dir) in
+    Command.run exe ~dir:(Path.build dir) ?stdout_to args
+  in
+  Super_context.add_rule sctx ~dir build
+
+let build_c_program ~sctx ~dir ~source_files ~scope ~cflags_txt ~output () =
+  let ctx = Super_context.context sctx in
+  let exe =
+    Ocaml_config.c_compiler ctx.Context.ocaml_config
+    |> Super_context.resolve_program ~loc:None ~dir sctx
+  in
+  let include_args =
+    (* XXX: need glob dependency *)
+    let ocaml_where = Path.to_string ctx.Context.stdlib_dir in
+    (* XXX: need glob dependency *)
+    let ctypes_include_dirs =
+      let ctypes = Lib_name.of_string "ctypes" in
+      let lib =
+        match Lib.DB.resolve (Scope.libs scope) (Loc.none, ctypes) with
+        | Ok lib -> lib
+        | Error _res ->
+          (* XXX: User_error.raise *)
+          failwith "error resolving 'ctypes' lib"
+      in
+      Lib.L.include_paths [lib]
+      |> Path.Set.to_list
+      |> List.map ~f:Path.to_string
+    in
+    let include_dirs = ocaml_where :: ctypes_include_dirs in
+    List.concat_map include_dirs ~f:(fun dir -> ["-I"; dir])
```

Comment:
I think it will be easier if you stick to `Command.Args.t` for representing arguments. Then you can do something like:

```ocaml
let pred = Predicate.create ~id ~f:(fun fn -> Filename.check_suffix ".h") in
List.map include_dirs ~f:(fun dir ->
  S [ Hidden_deps (Dep.Set.singleton (File_selector.create ~dir pred))
      ; A "-I"
      ; Path dir
     ]
```

Reply:
Hmm?   The purpose of this code is to generate these flags.

```
/usr/bin/gcc -I/usr/include -I /home/mbac/.opam/4.08.1+flambda/lib/ocaml -I /home/mbac/.opam/4.08.1+flambda/lib/ctypes
```

Not sure what you mean by the `Predicate.create ... Filename.check_suffix ".h"`